aten/src/ATen/cpp_custom_type_hack.h

```diff
@@ -26,6 +26,9 @@ T& cast(const Tensor& packed) {
 
 template <typename T>
 Tensor create(std::unique_ptr<T> ptr, TensorOptions options) {
+  // None of this should trace, so turn off Variable handling
+  at::AutoNonVariableTypeMode guard;
```

Comment:
umm...errr...I guess this is right?  But the tracing seems like a bit of an implementation detail, i.e. what you really want is some torch-level VariableType thing that does the "unpacking / setting the guards".  Which suggests this is at the wrong level?

Put another way -- I'm assuming because you put this here it was necessary, but what's the backtrace that makes this happen?

Reply:
From PR description

> The most subtle part of this patch is retaining tracing behavior: the fact that everything is a Variable means that more code gets routed to VariableType than before; this can change traces. I identified two places where we didn't appropriately turn off VariableType, mostly factory functions:

> * torch.tensor must turn off VariableType before invoking at::empty to construct the tensor, as it subsequently does direct data access
> * tensor_slow (invoked when you pass a Python scalar to a tensor argument) must turn off VariableType before calling scalar_to_tensor so the scalar gets traced as constant, rather than as a call to scalar_to_tensor.

I believe, concretely, for this case, quantized tests that utilized (1) tracing and (2) custom cpp type hack failed without this guard, because without the guard we spew an unnecessary `at::empty` into the trace.

Reply:
I'm a bit confused with the wording -- this is a third case?

Reply:
Yeah, it looks like I forgot to update the PR description with more cases as I found them.